# Data exploration and cleaning

Welcome to your first jupyter notebook! The first thing to know about Jupyter notebooks is that there are two kinds of cells. This is a markdown cell.

There are a lot of different ways to mark up the text in markdown cells, including __bold__ and *italics*.

The next one will be a `code` cell.

In [1]:
import pandas as pd

In [2]:
df = pd.read_excel('data_cards.xlsx')

## Verificación de la integridad de datos

Es vital entender qué significa cada fila y columna de un dataset. En este caso, se trata del nivel de crédito y datos financieros, cad fila representa una cuenta bancaria.

También se necesita elaborar o recurrir al diccionario de datos que es suna descripción del dataset de estudio, en este se detallan sus características y contexto. 

In [3]:
df.shape

(30000, 25)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          30000 non-null  object
 1   LIMIT_BAL                   30000 non-null  int64 
 2   SEX                         30000 non-null  int64 
 3   EDUCATION                   30000 non-null  int64 
 4   MARRIAGE                    30000 non-null  int64 
 5   AGE                         30000 non-null  int64 
 6   PAY_1                       30000 non-null  object
 7   PAY_2                       30000 non-null  int64 
 8   PAY_3                       30000 non-null  int64 
 9   PAY_4                       30000 non-null  int64 
 10  PAY_5                       30000 non-null  int64 
 11  PAY_6                       30000 non-null  int64 
 12  BILL_AMT1                   30000 non-null  int64 
 13  BILL_AMT2                   30000 non-null  in

In [5]:
df.columns.values.tolist()

['ID',
 'LIMIT_BAL',
 'SEX',
 'EDUCATION',
 'MARRIAGE',
 'AGE',
 'PAY_1',
 'PAY_2',
 'PAY_3',
 'PAY_4',
 'PAY_5',
 'PAY_6',
 'BILL_AMT1',
 'BILL_AMT2',
 'BILL_AMT3',
 'BILL_AMT4',
 'BILL_AMT5',
 'BILL_AMT6',
 'PAY_AMT1',
 'PAY_AMT2',
 'PAY_AMT3',
 'PAY_AMT4',
 'PAY_AMT5',
 'PAY_AMT6',
 'default payment next month']

In [6]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

Ahora verificamos si la columna ID tiene tantos valores únicos como filas en el dataset. El método "nunique()" hace un conteo de los valores únicos en el dataset y, como se puede ver, la cantidad de filas es mayor. Estosignifica que existe valoresduplicados en la columna **ID**

In [7]:
df['ID'].nunique()

29687

Ahora podemos usar el método **value_counts()** para hacer una lista de los valores únicos en la columna **ID** junto a la cantidad de veces que se repite cda uno de ellos.

In [8]:
id_counts = df['ID'].value_counts()
id_counts.head(10)

ad23fe5c-7b09    2
1fb3e3e6-a68d    2
89f8f447-fca8    2
7c9b7473-cc2f    2
90330d02-82d9    2
2a793ecf-05c6    2
75938fec-e5ec    2
7be61027-a493    2
a3a5c0fc-fdd6    2
b44b81b2-7789    2
Name: ID, dtype: int64

En efecto, existen filas con **ID** repetidos, pero si aplicamos la función value_counts() nuevamente

In [9]:
type(id_counts)

pandas.core.series.Series

De los 29687 ID identificados, 29374 solo están presentes una vez, en cambio 313 se repiten 2 veces. Nótese que no existen casos con un ID que se repite más de 2 veces.

In [10]:
id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

Resumiendo el proceso anterior...

La función values_counts se encarga de contar las veces que un valor se repite a lo largo de toda la columna. Luego creamos una máscar booleana (boolean mask) para filtrar los valores que se repiten 2 veces. Como es de esperar, la variable dupe_mask y id_counts tienen la misma longitud.

In [2]:
df = pd.read_excel('../Exploratory_Analysis/data_cards.xlsx')
id_counts = df['ID'].value_counts()
id_counts.head()
dupe_mask = id_counts == 2
dupe_mask[0:5]

ad23fe5c-7b09    True
1fb3e3e6-a68d    True
89f8f447-fca8    True
7c9b7473-cc2f    True
90330d02-82d9    True
Name: ID, dtype: bool

El resultado que tenemos hasta ahora es una dataframe con una columna que se identifica con el "ID" de la tabla original. A este tipo de estructura también se conoce como "series" de pandas.

In [26]:
type(dupe_mask)

pandas.core.series.Series

Dado que se requiere trabajar solo con las posiciones del ínide, estos serán apartados con el método index

In [3]:
id_counts.index[:5]

Index(['ad23fe5c-7b09', '1fb3e3e6-a68d', '89f8f447-fca8', '7c9b7473-cc2f',
       '90330d02-82d9'],
      dtype='object')

Existen 313 "ID" repetidos

In [5]:
# Aplicación del filtro
dupe_ids = id_counts.index[dupe_mask]
# Transformación a una lista
dupe_ids = list(dupe_ids)
# Consulta de la longitud de la lista
len(dupe_ids)

313

Verificación de los resultados

In [6]:
dupe_ids[:5]

['ad23fe5c-7b09',
 '1fb3e3e6-a68d',
 '89f8f447-fca8',
 '7c9b7473-cc2f',
 '90330d02-82d9']

Filtrando los primeros tres valores duplicados en el dataset

In [12]:
df.loc[df['ID'].isin(dupe_ids[:3]),:]

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
5033   89f8f447-fca8     320000    2          2         1   32     0      0   
5133   89f8f447-fca8          0    0          0         0    0     0      0   
16727  1fb3e3e6-a68d      80000    1          2         2   33     2      2   
16827  1fb3e3e6-a68d          0    0          0         0    0     0      0   
29685  ad23fe5c-7b09      50000    1          3         1   32     0      0   
29785  ad23fe5c-7b09          0    0          0         0    0     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
5033       0      0  ...     169371     172868     150827      8000      8000   
5133       0      0  ...          0          0          0         0         0   
16727      0      0  ...      27394      29922      31879         0      2000   
16827      0      0  ...          0          0          0         0         0   
29685      0      0  ...      12882       8131       3983      3000      2871   
29785      0      0  ...          0          0          0         0         0   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
5033       5500      6100      6000      5000                           0  
5133          0         0         0         0                           0  
16727      2000      3000      2600         0                           1  
16827         0         0         0         0                           0  
29685      1000       163      3983      3771                           1  
29785         0         0         0         0                           0  

[6 rows x 25 columns]

Por otro lado, si se pretende hallar todos los registro repetidos, no será necesario adicionar segmentaciones.

In [14]:
df.loc[df['ID'].isin(dupe_ids),:].head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
141  26bde6da-f148     180000    1          2         2   28    -1     -1   
196  42009c72-651f     150000    2          2         1   34    -2     -2   
241  26bde6da-f148          0    0          0         0    0     0      0   
296  42009c72-651f          0    0          0         0    0     0      0   
323  73ed1381-3a3f     230000    2          2         1   28     0      0   

     PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
141     -1     -1  ...        332        416        416         0       416   
196     -2     -2  ...        116          0       1500         0         0   
241      0      0  ...          0          0          0         0         0   
296      0      0  ...          0          0          0         0         0   
323      0     -2  ...       2100        767       6173      1710      1100   

     PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
141       332       500      3500       832                           0  
196       116         0      1500         0                           0  
241         0         0         0         0                           0  
296         0         0         0         0                           0  
323      1000       767      6173      1000                           1  

[5 rows x 25 columns]

One approach to deal with this issue would be to find rows that have all zeros, except for the first column, which has the IDs. These would be invalid data in any case, and it may be that if we get rid of all of these, we would also solve our problem of duplicate IDs. We can find the entries of the DataFrame that are equal to zero by creating a Boolean matrix that is the same size as the whole DataFrame, based on the "is equal to zero" condition.

In [15]:
df_zero_mask = df == 0

In [16]:
feature_zero_mask = df_zero_mask.iloc[:,1:].all(axis=1)

In [18]:
sum(feature_zero_mask)

315

El cálculo anterior nos dice que existen 315 filas que contienen 0 en todas columnas excepto la primera. Ya que es más grande que las filas duplicadas (313), podríamos eliminarlas solucionando el problema de los registros duplicados.

In [23]:
df_clean_1 = df.loc[~feature_zero_mask,:]

In [24]:
df_clean_1.shape

(29685, 25)

In [26]:
df_clean_1['ID'].nunique()

29685

Con el problema solucuonado, se procede a guardar el resultado como un archivo .csv

In [27]:
df_clean_1.to_csv('../Exploratory_Analysis/data_cards.csv', index = False)

Evaluación de los resultación

In [28]:
df_clean_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29685 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          29685 non-null  object
 1   LIMIT_BAL                   29685 non-null  int64 
 2   SEX                         29685 non-null  int64 
 3   EDUCATION                   29685 non-null  int64 
 4   MARRIAGE                    29685 non-null  int64 
 5   AGE                         29685 non-null  int64 
 6   PAY_1                       29685 non-null  object
 7   PAY_2                       29685 non-null  int64 
 8   PAY_3                       29685 non-null  int64 
 9   PAY_4                       29685 non-null  int64 
 10  PAY_5                       29685 non-null  int64 
 11  PAY_6                       29685 non-null  int64 
 12  BILL_AMT1                   29685 non-null  int64 
 13  BILL_AMT2                   29685 non-null  in

In [29]:
df_clean_1['PAY_1'].head(5)

0     2
1    -1
2     0
3     0
4    -1
Name: PAY_1, dtype: object

In [30]:
df_clean_1['PAY_1'].value_counts()

0                13087
-1                5047
1                 3261
Not available     3021
-2                2476
2                 2378
3                  292
4                   63
5                   23
8                   17
6                   11
7                    9
Name: PAY_1, dtype: int64

The preceding output reveals the presence of two undocumented values: 0 and –2, as well as the reason this column was imported by pandas as an object data type, instead of int64 as we would expect for integer data: there is a 'Not available' string present in this column, symbolizing missing data.

In [31]:
valid_pay_1_mask = df_clean_1['PAY_1'] != 'Not available'
valid_pay_1_mask[:5]

0    True
1    True
2    True
3    True
4    True
Name: PAY_1, dtype: bool

We see that 26,664 rows do not have the value 'Not available' in the PAY_1 column. We saw from the value count that 3,021 rows do have this value. Does this make sense? From Figure 1.23 we know there are 29,685 entries (rows) in the dataset, and 29,685 – 3,021 = 26,664, so this checks out.

In [32]:
sum(valid_pay_1_mask)

26664

In [35]:
df_clean_2 = df_clean_1.loc[valid_pay_1_mask,:]

In [37]:
df_clean_2.shape

(26664, 25)

In [39]:
df_clean_2['PAY_1'].value_counts()

 0    13087
-1     5047
 1     3261
-2     2476
 2     2378
 3      292
 4       63
 5       23
 8       17
 6       11
 7        9
Name: PAY_1, dtype: int64

Cambio de tipo de columna

In [40]:
df_clean_2['PAY_1'] = df_clean_2['PAY_1'].astype('int64')
df_clean_2[['PAY_1', 'PAY_2']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26664 entries, 0 to 29999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   PAY_1   26664 non-null  int64
 1   PAY_2   26664 non-null  int64
dtypes: int64(2)
memory usage: 624.9 KB


C:\Users\DELL\AppData\Local\Temp\ipykernel_11548\3119843425.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean_2['PAY_1'] = df_clean_2['PAY_1'].astype('int64')
